File: ForexConnectAPI-1.1.2-Linux-i686/samples/java/SimpleSamples/NonTableManagerSamples/GetHistPrices/src/ResponseListener.java
package gethistprices;

import com.fxcore2.IO2GResponseListener;
import com.fxcore2.O2GResponse;

public class ResponseListener implements IO2GResponseListener {
   
    //Response , log and error variables
    private SimpleLog mSimpleLog;
    private boolean mError = false;
    private O2GResponse mResponse = null;
    
    // Gets Response
    public O2GResponse getResponse() {
        return mResponse;
    }
    
    // Shows if there was an error during the request processing
    public final boolean hasError() {
        return mError;
    }
    
    // Constructor
    public ResponseListener(SimpleLog simpleLog) {
        mSimpleLog = simpleLog;
    }
    
    // Implementation of IO2GResponseListener interface public method onRequestCompleted
    public void onRequestCompleted(String requestID, O2GResponse response) {
        mResponse = response;
        mSimpleLog.log("Request {0} completed", requestID);
        mError = false;
    }

    // Implementation of IO2GResponseListener interface public method onRequestFailed
    public void onRequestFailed(String requestID, String error) {
        mSimpleLog.log("Request {0} failed. Error = {1}", requestID, error);
        mError = true;
    }

    // Implementation of IO2GResponseListener interface public method onTablesUpdates
    public void onTablesUpdates(O2GResponse response) {
    }   
    
}

File: ForexConnectAPIRobot/samples/cpp/SimpleSamples/NonTableManagerSamples/SubscriptionStatus/source/ResponseListener.cpp
#include "stdafx.h"
#include "SimpleLog.h"
#include "ResponseListener.h"

ResponseListener::ResponseListener(SimpleLog *log)
{
    mResponse = uni::CreateEvent(NULL, FALSE, FALSE, NULL);
    mRequestFailed = uni::CreateEvent(NULL, FALSE, FALSE, NULL);
    mLog = log;
    dwRef = 1;
    mLastResponse = NULL;
}

ResponseListener::~ResponseListener()
{
    uni::CloseHandle(mResponse);
    uni::CloseHandle(mRequestFailed);
    if (mLastResponse)
        mLastResponse->release();
}
/** */
void ResponseListener::setRequest(const char* requestID)
{
    mRequestID = requestID;
}
/** */
void ResponseListener::onRequestCompleted(const char * requestID, IO2GResponse *response)
{
    char buffer[1024];
    sprintf_s(buffer, sizeof(buffer), "Request completed requestID=%s", requestID);
    mLog->logString(buffer);
    if(mRequestID == requestID)
    {
        printf("Request completed\n");
        if (mLastResponse)
            mLastResponse->release();
        mLastResponse = response;
        mLastResponse ->addRef();
        uni::SetEvent(mResponse);
    }
}
IO2GResponse* ResponseListener::getLastRespnonse()
{
    mLastResponse->addRef();
    return mLastResponse;
}
void ResponseListener::onRequestFailed(const char *requestID, const char *error)
{
    char buffer[1024];
    sprintf_s(buffer, sizeof(buffer), "Request failed requestID=%s error=%s", requestID, error);
    mLog->logString(buffer);
    if (mRequestID == requestID)
    {
        printf("Request failed\n");
        uni::SetEvent(mRequestFailed);
    }
}

void ResponseListener::onTablesUpdates(IO2GResponse *data)
{
    mLog->addStringToLog("Data received");
    mLog->dumpLog();
    printf("Table updated\n");
    // uni::SetEvent(mResponse);
}

long ResponseListener::addRef()
{
    return InterlockedIncrement(&dwRef);
}

long ResponseListener::release()
{
    LONG dwRes = InterlockedDecrement(&dwRef);
    if (dwRes == 0)
        delete this;
    return dwRes;
}

